include/openthread/thread_ftd.h

```diff
@@ -649,10 +649,11 @@ otError otThreadSetMaxChildIpAddresses(otInstance *aInstance, uint8_t aMaxIpAddr
  */
 typedef enum
 {
-    OT_NEIGHBOR_TABLE_EVENT_CHILD_ADDED,    ///< A child is being added.
-    OT_NEIGHBOR_TABLE_EVENT_CHILD_REMOVED,  ///< A child is being removed.
-    OT_NEIGHBOR_TABLE_EVENT_ROUTER_ADDED,   ///< A router is being added.
-    OT_NEIGHBOR_TABLE_EVENT_ROUTER_REMOVED, ///< A router is being removed.
+    OT_NEIGHBOR_TABLE_EVENT_CHILD_ADDED,        ///< A child is being added.
+    OT_NEIGHBOR_TABLE_EVENT_CHILD_REMOVED,      ///< A child is being removed.
+    OT_NEIGHBOR_TABLE_EVENT_CHILD_MODE_CHANGED, ///< The child's device mode is changed.
```

Comment:
May not proper to add this event to `otNeighborTableEvent`?
I think `otNeighborTableEvent` are meant to notify neighbor table changes, meaning neighbor table has added or removed new entries. 

Maybe better to define it as a event of `Child`?
Thoughts?

Reply:
> Maybe better to define it as a event of Child?

Do you mean adding a new callback for "Child events"? 
My concern with this would be that it will add overhead of adding a new APIs/methods to register a new callback and also memory/RAM to save callback/context. 

`otNeighborTableEvent` is meant for changes in the neighbor table and we can consider child mode change as a neighbor table change (as if a child with old mode is removed and one with the new mode is added).

Reply:
> My concern with this would be that it will add overhead of adding a new APIs/methods to register a new callback and also memory/RAM to save callback/context.

We can add APIs/callbacks when that's required? For now, the event can only be handled in OT core.

Reply:
The main point for me is whether there is any advantage to add a new callback for "child mode change" over reusing the existing `otThreadRegisterNeighborTableCallback()`. I don't see any immediate advantage for adding new callback.

If we add the new "child mode change" event for internal OT core use, it does not add any overhead to add it in public `otNeighborTableEvent` as well.

The `otNeighborTableEvent` mechanism is intended to indicate changes in neighbor table which can also include signaling  changes to an existing child. 
- It was originally added in #2280 and later expanded in #3838.
- It was intended for two uses
   - (1) getting info on host fron NCP (e.g.in `wpantund`) for `StatCollector` and also logging (which was very helpful for debugging issue since we could see when different children attached/detached/re-atatched, etc., and 
   - (2) for telemetry use (again collecting info about child table changes). 
- I think signaling child mode change can be useful in both cases.

Reply:
SGTM to have one API to track all neighbor table changes.